File: Day2/Examples/C/type_contiguous.c
/*
 * 
 * A 1d ring example with MPI_Type_contiguous
 * 
*/
#include <stdio.h>
#include <mpi.h>

int main(int argc, char **argv){
  int i,ierror, myRank,uniSize;
  MPI_Status status;
  MPI_Datatype newtype;

  ierror=MPI_Init(&argc,&argv);
  ierror=MPI_Comm_size(MPI_COMM_WORLD,&uniSize);
  ierror=MPI_Comm_rank(MPI_COMM_WORLD,&myRank);

  int sBuf[2]={myRank, 0};
  int rBuf[2]={-1, 0};    

  int dest=(myRank < uniSize - 1) ? myRank + 1 : 0;
  int src=(myRank > 0) ? myRank - 1 : uniSize - 1;

  ierror=MPI_Type_contiguous(2, MPI_INT, &newtype);
  ierror=MPI_Type_commit(&newtype);

  while ( rBuf[0] != myRank ) {
    ierror=MPI_Sendrecv(sBuf, 1, newtype, dest, 100, rBuf, 1, newtype, src, 100,MPI_COMM_WORLD, &status);
    sBuf[1] += rBuf[0];
    sBuf[0] = rBuf[0];
  }

  printf("Rank %d, sums %d\n", myRank, sBuf[1] );
  ierror=MPI_Type_free(&newtype);

  ierror = MPI_Finalize();
  return ierror;
}

File: Day2/Examples/Fortran/type_contiguous.f90
!
! A 1d ring example with MPI_Type_contiguous
!
program typecontiguous
  use mpi_f08
  implicit none
  integer :: myRank, uniSize, ierror, i
  integer :: sbuf(2),rbuf(2), dest, src
  TYPE(MPI_Status) :: status
  TYPE(MPI_Datatype) :: newtype

  call MPI_Init(ierror)
  call MPI_Comm_size(MPI_COMM_WORLD,uniSize,ierror)
  call MPI_Comm_rank(MPI_COMM_WORLD,myRank,ierror)

  sBuf = (/myRank, 0/) 
  rBuf = (/-1, 0/)

  dest = modulo(myRank+1,uniSize)
  src  = modulo(myRank-1,uniSize)

  call MPI_Type_contiguous(2, MPI_INTEGER, newtype, ierror)
  call MPI_Type_commit(newtype, ierror)

  do while ( rBuf(1) .ne. myRank ) 
    call MPI_Sendrecv(sBuf, 1, newtype, dest, 100, rBuf, 1, newtype, src, 100, MPI_COMM_WORLD, status, ierror)
    sBuf(2) = sBuf(2) +  rBuf(1)
    sBuf(1) = rBuf(1)
  end do 
  
  print*, "Rank ", myRank, ", sum: ", sBuf(2)
  call MPI_Type_free(newtype, ierror)

  call MPI_Finalize(ierror)
end program typecontiguous
